doc/overall_design.md

```diff
@@ -0,0 +1,58 @@
+
+## Distributed Training
+
+![distributed_training_sequence](/doc/figures/distributed_training_sequence.jpg)
+
+## Dynamic Data Sharding
+
+**Elastic** is the key feature of ElasticDL. A worker can join and left at any time and the entire job still keeps running.
+
+The distributed execution of ElasticDL is data parallelism, not model parallelism. Each worker holds the whole model definition. Different shards of data are dispatched to different workers. Master doesn't care which worker reports the gradients, it just cares how many gradients are reported for the model version. In this way, adding or removing a worker won't interrupt the training process.
+
+At the start of an epoch, master node partitions the entire dataset into multiple shards which contain many records and then generates a todo list of tasks. Each task corresponds to a shard of data.\
+At the start point, each task is not assigned to any worker. The worker pulls a task (aka. a shard of data) at runtime and the master assigns the task to this worker. And then master moves this task to the doing list.\
+After processing this task and reporting the result, the worker will pull the next task and continue processing it.
+
+* If the task result is success, master removes the task from the *doing* list and mark it done.
+* If the task result is fail, master removes the task from doing list and then inserts it back into the *todo* list for recovery. Several issues can cause the task failure, such as the worker is preempted by a job of higher priority, the network connection is timeout and so on.
+
+![dynamic_data_sharding](/doc/figures/dynamic_data_sharding.png)
+
+The worker gets the task containing the data shard index (contains filename, startIndex, endIndex), it would be important to read the records of this shard efficiently from the data storage. In order to reach the IO efficiency, We choose the [RecordIO](https://github.com/elasticdl/recordio) data format for the input data.
+
+## Data IO Pipeline
+
+Data IO pipeline for elasticdl involve reading data from [RecordIO](https://github.com/elasticdl/recordio) file, making data generator for tf.data.Dataset and parsing features and label by dataset_fn user defined (see <em> Figure 2</em>).
+
+<center>
+    <img src="figures/data_io_pipeline.jpg" height="400" width="450">
+    <br>
+    <div style="
+    display: inline-block;
+    color: #999;
+    padding: 2px;"><em>Figure 2 </em>. elasticdl data IO pipeline</div>
+</center>
+
+After worker is launched, the worker will send request to get tasks from master. Each task contains the record index range [m, m+N) which can locate records in RecordIO file. DataReader read N records from RecordIO file by task index range and yield each record to create a generator. Then worker will perform the following steps to comsume record data from the generator:
```

Comment:
After worker is  -> After a worker is

Then worker -> Then the worker

Reply:
Good catch, updated.